### PR TITLE
operserv join rate warning thresholds

### LINKED ISSUE #49: port syn's joinrate.c
https://github.com/freenode/syn/blob/production/joinrate.c

should be easy to do as an atheme contrib module

### PR DESCRIPTION
**Pull request by jesopo**
a-la syn's `joinrate.c`. typical warning that i'm new to C. closes #49

### CODE REVIEW

File: src/os_joinrate.c [hunk 1]
```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 2021 notsurewhattoputhere
```
Comment (jesopo) [inline src/os_joinrate.c:2]: I don't know copyright law, but if I put a name for me here, can it just be "jess" with my email on it?

File: src/os_joinrate.c [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	if (joinrate->tokens > ch_rate)
+		joinrate->tokens = ch_rate;
+
+	// is there a token left for us to consume?
+	if (joinrate->tokens >= 1)
+	{
+		// yes, consume a token
+		joinrate->tokens--;
+	}
+	else if (CURRTIME - joinrate->last_warn_time >= 30)
```
Comment (jesopo) [inline src/os_joinrate.c:124]: would be nice to specify the `30` here in a setting

File: src/os_joinrate.c [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+	db_register_type_handler("JR", db_handle_jr);
+
+	service_named_bind_command("operserv", &os_joinrate);
+
+	os_joinrate_heap = mowgli_heap_create(sizeof(joinrate_t), 32, BH_LAZY);
+	os_joinrates = mowgli_patricia_create(irccasecanon);
+
+	joinrate_default = os_joinrate_create("default");
+	joinrate_default->rate = 5;
+	joinrate_default->time = 5;
```
Comment (jesopo) [inline src/os_joinrate.c:281]: would be nice to specify these defaults in settings

File: src/os_joinrate.c [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+	{
+		// yes, consume a token
+		joinrate->tokens--;
+	}
+	else if (CURRTIME - joinrate->last_warn_time >= 30)
+	{
+		// no, spit out a warning
+		joinrate->last_warn_time = CURRTIME;
+		slog(LG_INFO, "JOINRATE: \2%s\2 exceeds warning threshold (%d joins in %ds)",
+		cu->chan->name, joinrate->rate, joinrate->time);
```
Comment (skizzerz) [inline src/os_joinrate.c:129]:
```suggestion
		cu->chan->name, ch_rate, ch_time);
```

File: src/os_joinrate.c [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+	hook_add_db_write(db_write_jr);
+	db_register_type_handler("JR", db_handle_jr);
+
+	service_named_bind_command("operserv", &os_joinrate);
+
+	os_joinrate_heap = mowgli_heap_create(sizeof(joinrate_t), 32, BH_LAZY);
+	os_joinrates = mowgli_patricia_create(irccasecanon);
+
+	joinrate_default = os_joinrate_create("default");
```
[the remark this answers is not in the document]
Comment (aaronmdjones) [reply src/os_joinrate.c:279]: See above re: `os_joinrate_create()`

**Pull request closed without merging**